MudOverlay: Add parameter Modal allowing click-through

Cybrosys opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description

Introduces a new opt-in behavior that allows pointer events to pass through, enabling users to interact with other elements on the page directly. The overlay will close upon interaction with other elements if `AutoClose` is `true`, eliminating the need for an extra dismiss click/tap.

The Modal parameter has been introduced to several components that currently use `MudOverlay` to handle dismissal or toggling between open and closed states, such as menus and picker controls, this to achieve the expected behavior while not blocking interaction with elements behind the overlay.

https://github.com/MudBlazor/MudBlazor/issues/10738

Extra: #10938 

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

Unit and visually

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Cybrosys committed:
Add MudMenu class to listen for mousedown events and if they happen outside of any mud-menu-list it will invoke the CloseMenuAsync method

Cybrosys committed:
Add OverlayStyle to MudOverlay for dynamic styling

Cybrosys committed:
Add property Modal and update Open and Close methods to invoke the mousedown event listening methods

Cybrosys committed:
Update documentation

versile2 commented:
I thought we were talking about putting it on the Overlay so it would apply to all controls that use the overlay instead of just mudmenu

Cybrosys commented:
> I thought we were talking about putting it on the Overlay so it would apply to all controls that use the overlay instead of just mudmenu

Ah, my misunderstanding! I’ll look into shifting it to the Overlay.

Cybrosys commented:
I'm struggling to implement this in `MudOverlay` because using `pointer-events: none` prevents the overlay from being returned when using `elementsFromPoint()` or similar methods to determine which element is directly underneath the cursor.

In the current PR, the logic checks whether the clicked element is inside any `div` with the class `mud-menu-list`. I could likely query elements at a specific point as well. If the click occurs inside an element with this class, the menu remains open.

At this point, the only approach I see is for the component using `MudOverlay` to explicitly provide a class or identifier that `MudOverlay` can query and check against (for example, `MudMenu` would provide `mud-menu-list`). 

I'm open to any other suggestions or insights on how best to approach this. Let me know what you think!

versile2 commented:
so there are only a handful of popover / overlay combo controls (autocomplete, select, picker, menu?) so it's easy to look at each one. I think the only reasonable way I could think about it would be to 

a) create a css class on the overlay something like `mud-overlay-allow-clickthrough` then in mudPopoverJs where I'm adjusting the overlay css you have access to the popover element ID etc so it would be easy to ignore it on a click handler. If that class exists then do that, etc.

b) create a separate component and add Overlay true/false default to true on the components that use it, then anyone who wanted overlay functionality with clickthrough could use the new component after setting Overlay=false on the popover component. I'm unsure of this one but it's an idea that went through my head.

c) create a js item that is popover agnostic and accepts an element. In each of the popover/ overlay components add a property OverlayClickThrough that when true onafterrender will attach the element to the js function. On the overlay create a method that does that allowing the dev to pass the element to the Overlay. 

And before I did any of those I'd wait for @danielchalmers to chime in, he may hate/love one over the other or decide not to move forward with it. That way if you do the work at least your pretty sure it will be accepted.

Cybrosys commented:
The core issue is that once we use `pointer-events: none` to allow interaction through the overlay (and also show the correct mouse cursor), the overlay no longer receives any click events. JavaScript does not support forwarding click events or generating new ones.

The fourth solution d) would be to add a new css class `mud-overlay-block-clickthrough` on elements that should block passthrough. For the `MudMenu` this would be the mud menu items container. I would also add and remove the css class conditionally on the activator element. This would allow the `MudOverlay` JS to check if there are any elements under the pointer with the css class or not. If no elements have the class then it calls a method on the dotnetref (`MudOverlay`).

danielchalmers commented:
I apologize, I can't suggest a firm direction right now. The `mud-overlay-allow-clickthrough` example (solution a+d) seems logical since @versile2 has used that technique a few times, but I'll defer to him as he is experienced in this

danielchalmers commented:
I'll add that MUI is an interesting example: [Autocomplete](https://mui.com/material-ui/react-autocomplete/) is non-modal, but [Select](https://mui.com/material-ui/react-select/) and [Menu](https://mui.com/material-ui/react-menu/) are.

Cybrosys commented:
> I apologize, I can't suggest a firm direction right now. The `mud-overlay-allow-clickthrough` example (solution a+d) seems logical since @versile2 has used that technique a few times, but I'll defer to him as he is experienced in this

I'm not fully following how `mud-overlay-allow-clickthrough` should or could be used. As I mentioned, using `pointer-events: none` allows the cursor to interact with elements behind the `MudOverlay`. However, this also causes the overlay to be excluded from interaction events, meaning it won't be returned when querying elements under the cursor and it will no longer receive any click events.

What I've found so far is that a solution could focus on the element(s) containing the popup content. Using the `MudMenu` example, the root element of the menu items would serve as the element reference. A global click event would then check if the element under the cursor is a child of the menu’s root element. If it's not, the click is considered outside the popup, and the menu is dismissed.

In this context, `MudOverlay` seems to primarily serve the purpose of _preventing_ interaction with background elements (true modal behavior) and optionally shading the background.

That said, I might have misunderstood the intended usage of `mud-overlay-allow-clickthrough`, so I appreciate any further clarification.

Cybrosys commented:
> I apologize, I can't suggest a firm direction right now. The `mud-overlay-allow-clickthrough` example (solution a+d) seems logical since @versile2 has used that technique a few times, but I'll defer to him as he is experienced in this

Sorry, re-adding my comment. I thought I was onto something regarding how `MudOverlay` was currently preventing interaction but it was merely an effect of z-index and absolute positioning.

**Original comment:**
I'm not fully following how `mud-overlay-allow-clickthrough` should or could be used. As I mentioned, using `pointer-events: none` allows the cursor to interact with elements behind the `MudOverlay`. However, this also causes the overlay to be excluded from interaction events, meaning it won't be returned when querying elements under the cursor, and it will no longer receive any click events.

From what I've found through research and testing, a reliable solution should focus on the element(s) containing the popup content. Using the `MudMenu` example, the root element of the menu items would serve as the reference. A global click event would then check if the element under the cursor is a child of the menu’s root element. If it's not, the click is considered outside the popup, and the menu is dismissed. However, as that is not fully reliable I would still opt for the use of a CSS class for the _modal_ elements (menu items).

In this context, MudOverlay seems to primarily serve the purpose of preventing interaction with background elements (true modal behavior) and optionally shading the background.

That said, I might have misunderstood the intended usage of `mud-overlay-allow-clickthrough`, so I appreciate any further clarification.

versile2 commented:
Just a blind one eye pass on it from behind the keyboard not running the code.

so on the click handler of the mudoverlay it would check if that class exists if it does it would pass it's elementreference and event (the clientX and clientY of mouseeventargs) to a javascript function at the end of the click handler. So it completes the overlay purpose then runs the following javascript function.

focusElementUnderOverlay(element, event) {
  // make sure overlay valid
  if (!element) return;
  // get element underneath
  const elementUnder = document.elementFromPoint(event.clientX, event.clientY);

 // pass through
  if (elementUnder) {
    elementUnder.click(); // might should be focus?
  }
}

Cybrosys commented:
For that to technically work we would not be able to use `pointer-events: none`. If we are not using it then you won't see any visual feedback as the mouse goes over different elements, elements won't highlight and the cursor won't change.

If you run the docs server in this PR and see the updates docs example for the Menu, you can see how the cursor reacts and interacts with the elements differently depending on if the "Modal Menu" or "Non-modal Menu" is showing.

versile2 commented:
then the flip side is onafterrender of mudoverlay you remove the click handler if that class exists, run your js, at the click of your js execute the close handler of the mudoverlay from dotnetref. In your js if the item you click has the same or higher zindex then you do nothing. Again spitballing I'm not coding some or all of this might not work like we want.

Cybrosys commented:
> then the flip side is onafterrender of mudoverlay you remove the click handler if that class exists, run your js, at the click of your js execute the close handler of the mudoverlay from dotnetref. In your js if the item you click has the same or higher zindex then you do nothing. Again spitballing I'm not coding some or all of this might not work like we want.

I tried checking the z-index of the clicked element to try and compare it with the z-index of the mud overlay, to see if it's in front of it or not, but for some reason the clicked element's z-index was always lower than the overlay's. I might not have done it properly either, so might be worth a retry.

versile2 commented:
> > then the flip side is onafterrender of mudoverlay you remove the click handler if that class exists, run your js, at the click of your js execute the close handler of the mudoverlay from dotnetref. In your js if the item you click has the same or higher zindex then you do nothing. Again spitballing I'm not coding some or all of this might not work like we want.
> 
> I tried checking the z-index of the clicked element to try and compare it with the z-index of the mud overlay, to see if it's in front of it or not, but for some reason the clicked element's z-index was always lower than the overlay's. I might not have done it properly, so might be worth a retry.

aye let me know if you have a code set that's close I'll take a look. Might be some other markers too, if they are all popovers then traverse to see if data-ticks is greater than 0, if it is it's an active popover and should be ignored.

Cybrosys commented:
> > > then the flip side is onafterrender of mudoverlay you remove the click handler if that class exists, run your js, at the click of your js execute the close handler of the mudoverlay from dotnetref. In your js if the item you click has the same or higher zindex then you do nothing. Again spitballing I'm not coding some or all of this might not work like we want.
> > 
> > 
> > I tried checking the z-index of the clicked element to try and compare it with the z-index of the mud overlay, to see if it's in front of it or not, but for some reason the clicked element's z-index was always lower than the overlay's. I might not have done it properly, so might be worth a retry.
> 
> aye let me know if you have a code set that's close I'll take a look. Might be some other markers too, if they are all popovers then traverse to see if data-ticks is greater than 0, if it is it's an active popover and should be ignored.

I did a test and the z-index was not reliable. I found some people on SO who had tried to solve it but it escalated into someone writing code that mimics/builds the visual hierarchy of the entire page with JS, let's not use that.

Analyzing the DOM hierarchy, everything in front of the mud overlay will be a child under a popovercontent-div. The parent divs could be identified by the data-ticks attribute, but also the CSS classes `mud-popover` and maybe `mud-popover-open`. The mud overlay is at the same DOM depth as the popovercontent-divs.

We could solve it by either:
a) Calculating the DOM depth of the clicked element and mud overlay, and comparing them.
b) Traverse up the DOM tree from the clicked element until we find a div with data-ticks > 0 and/or one that has the CSS class(es) `mud-popover`

I'll try and prep a trymud or sample with a and b to choose between.

versile2 commented:
sounds good, you can use javascript on https://try.arctechonline.tech if needed

Cybrosys commented:
> sounds good, you can use javascript on https://try.arctechonline.tech if needed

Here's a thrown together [TryMudBlazor snippet](https://try.mudblazor.com/snippet/cOwJumQcTQIlZNLJ) based on some of my findings and tests.

### Key Findings:
- Using DOM depth to determine if an element is behind the overlay was unreliable. Elements behind the overlay could have varying DOM depths depending on page content, making it unrelated to the overlay's depth.
- The sample includes:
  1. A **modal menu**  
  2. A **popover toggle button**  
  3. A **non-modal menu** (with `pointer-events: none` on the overlay)  

- The popover demonstrates how components using `MudPopoverProvider` can produce false positives.  
- `MudDialog` doesn’t use `MudPopoverProvider`, which is worth noting.

### JavaScript Methods:
1. **`isTargetBehindOverlay`**:  
   - Checks if the clicked target (or its ancestor) shares the same parent as `MudOverlay`.  
   - Works for both components using `MudPopoverProvider` and those that don’t, like `MudDialog`.

2. **`isTargetPopoverContent`** (less reliable):  
   - Requires `MudPopoverProvider`.  
   - Uses `element.closest()` to check if the element or its ancestors have the `mud-popover-open` CSS class.

### Additional Suggestion:
- Another possible approach would be to introduce new CSS classes specifically for identifying elements that should not cause the overlay to be dismissed, providing more granular control when a non-modal behavior is desired.
- In my local branch I am controlling the CSS style on the `MudOverlay` using a parameter called `Modal` that is also exposed on other components such as the `MudMenu` (`<MudMenu Modal="false">`).

versile2 commented:
ok so looks good and what is required to make this a fairly generic thing for overlay? Such as an option on Overlay for Modal true/false or do we have to add a property on everything that uses an Overlay. Walk me through it, but I like it and apart from some testing I can't see it not working right.

Cybrosys commented:
The term `Modal` is well-known in app development, referring to an element that appears in front of others and blocks interaction with the background. As such I don't feel that `Modal = true/false` is a bad choice for the `MudOverlay`. If we look at some other parameters that _activate_ features, maybe the parameter should be called `Modeless`.

I believe that the current default behavior (`Modal = true`) has advantages for touch-based interfaces. So whether this should be the default behavior is worth discussing with a broader audience, as there may be unexpected side effects or differing opinions.

If we keep to having the current behavior the default, so this would be "opt-in". I would then suggest adding the same parameter (`Modal`) to controls that use the `MudOverlay` for dismissing popup elements and bind it. This means that to be able to interact with the background while a `MudMenu` is showing would look something this:
```
<MudMenu Modal="false">...</MudMenu>
```

This feature also ties in to `AutoClose` on `MudOverlay`. So the requirement for activating the JavaScript listener is that `AutoClose` is `true` and `Modal` is `false`.

Cybrosys commented:
Maybe add `Modeless` or `Modal` to `MudGlobal` so it's possible to opt-in globally for the new behavior?

versile2 commented:
We won't purposefully introduce a breaking change outside of major releases so we must keep current behavior or shelve this until we are closer to V9.
---
I like Modal true/false over Anything else but I also don't like having to put it on each item that uses Overlay. Also exclude Dialogs and Drawer. Those aren't our targets for this, select, autocomplete, menu are the top 3. And some way to trigger from the Overlay itself? public method SetModelessElement(ElementReference element) or something like that?

Cybrosys commented:
OK. Let's go with `Modal` for now. I will update the branch and then we can continue on how to improve it from there.

Cybrosys committed:
Move solution to MudOverlay

Cybrosys committed:
Merge branch 'MudBlazor:dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Adjust whitespace

Cybrosys committed:
Satisfy SonarQube by assigning the value to a dummy variable

Cybrosys committed:
Use elementFromPoint instead of elementsFromPoint

Cybrosys commented:
I've updated the branch so you can see the current code changes.

> And some way to trigger from the Overlay itself? public method SetModelessElement(ElementReference element) or something like that?

Could you clarify this the above, I interpreted it as wanting to pass the element reference to the JavaScript?

I did a test passing in the div element as a reference to the JavaScript. It added extra complexity as I had to check in `OnAfterRenderAsync` when an `ElementReference?` had been given a value from the `@ref` as the `RenderFragment` is conditionally rendered.

The current checked in code relies on finding an element with the `mud-overlay` CSS class instead of passing in an element reference.

One limitation that I'm unsure of how to solve is being able to support multiple non-modal overlays with `AutoClose`. The current script only supports one. One idea is to support passing in multiple references and close all the overlays that are under the cursor as long as there is no other element in between.

versile2 commented:
ok so I won't have time to delve into this for a few hours likely but - the overlay is a Section component which means unless it's excluded (Dialog, Drawer, Absolute) there will only ever be one and it will always be the first div under mud popover provider.

Cybrosys commented:
The question is whether to document and accept the limitation for now or to support multiple modeless overlays.

If we choose to support multiple overlays, I suggest the following approach:  
- Maintain a map (or list) in JavaScript of all visible modeless overlays with auto-close enabled.  
- Since neither elements nor DotNet references can be used as map keys, each overlay should be assigned a unique ID.

On `mousedown`, the JavaScript would:  
1. Iterate over all overlays in the map and temporarily update their styles.  
2. Trigger a reflow.  
3. Use `elementsFromPoint()` to determine all elements at the mouse position.  
4. Restore the original styles.  
5. Check if the top-most element is an overlay and whether any additional overlays are directly underneath it.  
6. If one or more overlays are detected, invoke their `CloseOverlayAsync` methods.  
7. The JavaScript can remove the overlays themselves from the map directly or rely on a callback from the component to unregister (`cancelListener(id)`).

This approach would enable support for multiple modeless overlays. I have some concerns about potential memory leaks if overlays are not properly removed from the map. To mitigate this, we could add an additional call in `Dispose` to ensure that the overlay is removed from the map if it exists.

An alternative solution could be:
On `mousedown`, the JavaScript would:
1. Find all elements with the CSS class `mud-overlay` and `pointer-events: none`.
2. Temporarily update their style to `pointer-events: auto`.
3. Use `elementsFromPoint()` to get all elements at the mouse position.
4. Restore the original `pointer-events: none` style.
5. If the top-most element is a `mud-overlay` and other `mud-overlay` elements are directly underneath it:
   - Build a list of those overlays and their `id` attributes.
6. Invoke a C# method (`CheckOverlaysAsync`) and pass the list of overlay IDs.

Each C# overlay component will compare the received IDs with its own `UniqueId` and call `CloseOverlayAsync` if necessary.

Cybrosys commented:
Since both of the above solutions would require a unique ID, we could bind it directly to the overlay `<div>`. This way, we can check element IDs instead of passing `ElementReference` instances around.

Cybrosys committed:
Add support for multiple modeless overlays

Cybrosys committed:
Use `pointerdown` instead of `mousedown` for broader input support

Cybrosys committed:
Initial support for MudSelect and fix ES6 compatibility

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

versile2 commented on src/MudBlazor/Components/Overlay/MudOverlay.razor.cs:
```diff
@@ -196,17 +220,44 @@ protected override async Task OnAfterRenderAsync(bool firstTime)
         }
     }
 
+    protected async Task OnVisibleChanged(ParameterChangedEventArgs<bool> args)
```
OnVisibleChanged fired if Overlay Visible="true" when it's first shown?

Cybrosys replied:
Good catch, I'll do a sanity test. It should call the event as the as the parameter registration happens in the constructor before the value is set from Razor. If the Visible property had a default value of true, then I think it wouldn't fire.

I could add a check in `OnAfterRender` when `firstTime` is `true` as well.

versile2 replied:
What I would do (might check other components see how they handle it (not the popover)) is have a flag that indicates it is loaded. OnParameterSet if that flag isn't set/true/whatever set it. I believe that fires every time a parameter is set even the first.

versile2 commented on src/MudBlazor/Components/Select/MudSelect.razor:
```diff
@@ -95,5 +97,19 @@
-<!-- pointerdown instead of click needed to close the menu before OnLostFocus runs -->
-<MudOverlay Visible="_open" @onpointerdown="@(() => CloseMenu(false))" LockScroll="@LockScroll" />
\ No newline at end of file
+@if (Modal)
```
I feel like there is someway we could use just one Overlay with a Model parameter to be toggled instead of an if one version of overlay else another version of overlay. Also onpointerdown can replace the click handler in AutoClose so you can accomplish this? Is there a downside to that?

Cybrosys replied:
Yes, I would like that as well. This was a quick first initial support so I could update the docs example and test. I need to look into the comment regarding `OnLostFocus` and why it was important to close the menu before it fired. There seems to be a desired chain of events there that I don't want to break.

versile2 replied:
Apart from your overlay changes I would try and leave the component itself alone. Odds are the overlay change won't affect what they are doing.

Cybrosys replied:
I updated the `MudSelect` so it now only has one overlay and have the same behavior regardless if it's modal or modeless.

Regarding updating the `AutoClose` feature to react on `onpointerdown` instead of `onclick`. One undesired effect could be that unless we add extra checks/filtering, it would close when right-clicking as well. It should be possible to check the event if it's from a mouse and filter on which button. Hard for me to say if it would be better, worse or the same.

Cybrosys replied:
After sleeping on it, I will say that we can close on `@onpointerdown` but we should keep `@onclick:stopPropagation` to avoid introducing any breaking changes.

There are components such as the `MudDialog` that checks `OnClick` on the `MudOverlay` to check if the dialog can be closed by clicking outside of it. The dialog has essentially reimplemented `AutoClose`. The question then is if the dialog should be updated to rely on `AutoClose` as well or not. We also have to consider if `OnClick` should be kept or if we should change it to `OnPointerDown`.

I would like to avoid having a mix of auto-closing behaviors where some components close on click events and others on pointer down events. The reason why the new JS reacts to pointer down is due to the inherent behavior of modeless where I want the overlay gone as fast as possible so it doe not interfere with interactions against controls in the background.

If you want `onpointerdown` to be the primary event for `AutoClose`, then I would suggest that we also update other components and their logic to rely on `AutoClose` instead of re-implementing it, like `MudDialog` has. The risk is that this PR will grow a bit more.

Cybrosys replied:
After further consideration, I believe the choice between reacting to `onpointerdown` vs. `onclick` depends on the use case.

For modal dialogs that block interaction with background elements, requiring a full click is expected to prevent accidental dismissals (swipe gestures etc.). However, when the overlay is used for dismissing UI elements (e.g., dropdowns or popovers), `onpointerdown` is often the better choice for a more responsive experience.

I believe that `MudOverlay`' s current usage of `onclick` works when `Modal` is `true`. I would suggest that for the next major version, each component is configured per use case to have their default `Modal` value either kept as `true` or set to `false`.

To give some examples:
The `MudDialog` when dismissal should happen when clicking/tapping outside, should be `Modal = true`.
The `MudSelect` who use the overlay for dismissing their items should have `Modal = false`.

versile2 reviewed (changes requested):
Overall really good job and I'm pretty happy. I'd like to look at the two comments I made if they can be improved. Stellar job, just want it to be as smooth as possible with no impact to existing stuff. Any reason you didn't add Autocomplete in as well?

Cybrosys committed:
Update MudSelect to use the same MudOverlay regardless of if Modal or Modeless. Updated the comment to explain the parameter bindings.

Cybrosys committed:
Fixed an issue where the overlay's modeless auto-close tracking wouldn't be registered if initially visible

Cybrosys committed:
Syntactic refactor to see if codecov stops complaining

versile2 commented:
The unit tests are as much designed to ensure there are no obvious mistakes as much to prevent unnecessary breaking changes so you will need the following unit tests.
	One for MudOverlay to verify id is assigned
	One for MudOverlay to verify OnVisibleChanged is called on init, when changed, and if you need to verify dispose
	One for MudOverlay to verify what happens when CloseOverlayAsync is called
	Double check any methods that don't have coverage but I think that's it
	If you want no code quality you can mock the javascript stuff.
	A quick test on each component that is using Modal to ensure setting Modal true/false has appropriate changes
	
If you have any trouble with tests or want to chat pm me on Discord. Request review again when ready for me to re-review. It might even be cleaner to make a new PR and close this one so the other reviewers won't have to read 100 messages hehe but that's not a biggie.

Cybrosys committed:
Add OverlayPointerDownListener and update MudOverlay to use it

Cybrosys committed:
Remove comment on overlay closure behavior

Cybrosys committed:
Add unit tests to verify that create and start is called when expected

Cybrosys committed:
Start and stop listening to pointerdown in OnParametersSetAsync to be sure any changes to modal, visible and autoclose results in expected behavior.

Update OverlayPointerDownListener to track if listening to skip calling JS interop when not needed.

Cybrosys committed:
Fix whitespace

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Add modal to MudPicker

Cybrosys committed:
Latest from dev

Cybrosys committed:
Add unit tests for OverlayPointerDownListener and Factory

Cybrosys committed:
Remove redundant whitespace

Cybrosys committed:
Add unit test to verify auto close when IOverlayPointerDownListener.OnPointerDown is invoked

Cybrosys committed:
Update color picker documentation

Cybrosys committed:
Update documentation and examples for DatePicker, DateRangePicker and TimePicker

Cybrosys committed:
Let's not forget Autocomplete

Cybrosys committed:
Undo fix for Appbar until confirmation

Cybrosys committed:
Set modal to false on the autocomplete element to allow dismissal of the dialog when clicking outside

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

versile2 commented:
I will try to review tomorrow, thanks for following through with this and the ever changing target.

Cybrosys commented:
> I will try to review tomorrow, thanks for following through with this and the ever changing target.

Of course! I wanted this feature implemented as well. Let me know if you have any questions, especially regarding the JavaScript and the surrounding service. In theory, it could be used generically by any element, but I kept it named to the overlay since there's no current need for broader use. There are also some specifics in it that would need to be controlled via settings to make it fully generic.

ScarletKuro commented:
I'm not sure I like all this complexity and additional JS for something so specific that solves only one problem. It feels like we've stopped trying to innovate and find creative ways to avoid JS. 
From my understanding, this service essentially just listens for `pointerdown`? Could we not reuse the `IEventListener` that was created for similar purposes, or would it require fine-tuning to work specifically with the overlay? 
I'm also concerned that `OverlayPointerDownListenerFactory` doesn't meet the new requirements. I understand it was done by analogue of `IResizeObserver`, `IScrollSpy`, etc., but we no longer want to use these patterns. The old service, which relies on `event` which doesn't support async code, and we don't want to continue using these factories and transient services that create a 1:1 connection between a service and a component. We'd rather have a 1:many relationship. 
A good example of the new approach can be found in `KeyInterceptorService` and `BrowserViewportService`. These support both sync and async callbacks, and a single service can handle multiple subscriptions. The API is generally better, and these are the new services that have replaced the old pattern currently used by OP.

Cybrosys commented:
> From my understanding, this service essentially just listens for `pointerdown`? Could we not reuse the `IEventListener` that was created for similar purposes, or would it require fine-tuning to work specifically with the overlay?

Using `pointer-events: none` removes the element from receiving any pointer events, which also means it is excluded from JavaScript methods like `elementsFromPoint`. The only way to query the element again is to temporarily change `pointer-events` back to `auto`, call `elementsFromPoint`, and then reset it to `none`.

Calling `elementsFromPoint` is crucial here because we only want to react if the overlay is the topmost element under the pointer. If we were to use `IEventListener`, we would have to listen for the event, update the element’s style from Blazor C#, call `elementsFromPoint`, evaluate if it’s the topmost element, and then restore the style. This would or could result in multiple JavaScript-to-C# roundtrips.

But I might have misunderstood or missed something with how we could make use if `IEventListener`. So to be able to react as fast as possible, I wrote it as an isolated service. I am not familiar with the full Blazor source code and its history.

> I'm also concerned that OverlayPointerDownListenerFactory doesn't meet the new requirements. I understand it was done by analogue of IResizeObserver, IScrollSpy, etc., but we no longer want to use these patterns. The old service, which relies on event doesn't support async code, and we don't want to continue using these factories and transient services that create a 1:1 connection between a service and a component. We'd rather have a 1:many relationship.
A good example of the new approach can be found in KeyInterceptorService and BrowserViewportService.

I’ll have to investigate the pattern that `KeyInterceptorService` and `BrowserViewportService` implement. I chose the factory variant because it seemed simpler.

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

ScarletKuro commented:
> But I might have misunderstood or missed something with how we could make use of `IEventListener`.

It was just a question of whether we could reuse the existing logic. From your explanation, I now see that we unfortunately can't.

Whats @danielchalmers @henon opinion on this? The problem itself is simple and clear, however the implementation adds additional complexity, a new parameter and new very specific service.

henon commented:
My opinion is that the functionality is well worth the added complexity. If I understand the problem correctly we want to be able to click through the (invisible) overlay to directly start interacting with other elements on the page and at the same time close the open menu avoiding a second click. This is how most would expect menus to work so our menus should be able to do it. 

As for how the service that does it should be implemented in order to make this functionality at least reusable should we need it again I'll adhere to @ScarletKuro 's advice.

ScarletKuro commented:
> If I understand the problem correctly we want to be able to click through the (invisible) overlay to directly start interacting with other elements on the page and at the same time close the open menu avoiding a second click

My understanding is the same, which means that the root of the issue is still the popover used in MudMenu in conjunction with MudOverlay. This makes me want to reconsider my idea of abandoning the integration of Floating UI, which theoretically wouldn't require all these overlay 'tricks' to close the menu.

Cybrosys commented:
This work initially started with `MudMenu`, but after a request to implement it from the perspective of `MudOverlay` and the `AutoClose` parameter/feature, it became evident that several other controls with open/closed states could also benefit from this enhancement.

This PR introduces the `Modal` parameter to the following controls, ensuring they exhibit the expected behavior of allowing interaction with other elements on the page while remaining in their open state. The docs have been updated for the various components to experience the different Modal behaviors:

- Autocomplete
- ColorPicker
- DatePicker
- DateRangePicker
- Menu
- Select
- TimePicker

henon commented:
> The default value for `Modal` is `true`, which is the current behavior where the overlay blocks interaction.

We should consider changing this default value to `false` in v9. I can't even think of a reason why you wouldn't want to directly interact with the next input component after having selected a value in Menu, Autocomplete, Select and all the pickers. 

The `Modal` behavior is really something I only know from dialogs and message boxes

Cybrosys commented:
Changing the default `Modal` value from `true` to `false` raises a potential concern: will controls close properly when focus shifts via the Tab key? I haven’t looked into this yet, but some controls may already handle it through the blur event. It’s likely not a major issue, but worth keeping in mind. I don't know if the overlay today blocks the use of the Tab key or not.

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys commented:
> A good example of the new approach can be found in `KeyInterceptorService` and `BrowserViewportService`. These support both sync and async callbacks, and a single service can handle multiple subscriptions.

@ScarletKuro Are you okay with keeping the service somewhat tied to the overlay in namespace and naming, calling it `IOverlayPointerDownService`?

It could be made more generic to focus on the `pointer-events: none` scenario and configuration for elements, but I’m not sure what the best name would be in that case. That could always be revisited when the need arises, likely with additional requirements that aren’t clear at this point.

danielchalmers commented:
Just a note: Some components may not work as expected if you try to click them through their own overlay (ex: Autocomplete/Select). It might toggle the overlay or only open and not close again depending on the assumption they make that they could never be in that state (haven't checked those specifically but you see my point).

The feature is valuable and makes things feel more fluid. I agree we should explore making it the default at some point. For the implementation, like henon, I'll look to ScarletKuro.

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

versile2 commented:
> I'm am currently testing the requested changes and I'm experiencing a small undesired behavior in `MudMenu` as a result of the changes in #10853 (the `MudPopover` and `MudOverlay` is conditionally rendered).
> 
> This conditional rendering is preventing expected events in the `MudOverlay` from firing. For example `OnParametersSetAsync` is not called when it no longer is visible. The `ParameterState<bool> _visibleState` also misbehaves.
> 
> If I move the overlay outside of the conditional rendering block, everything works, does anyone have any objections to me doing this change?
> 
> ```cs
> @* The portal has to include the cascading values inside because it's not able to teletransport the cascade *@
> @if (_openState.Value)
> {
>     <MudPopover Open="@_openState.Value"
>                 Class="@PopoverClassname"
>                 Style="@Stylename"
>                 MaxHeight="@MaxHeight"
>                 AnchorOrigin="@GetAnchorOrigin()"
>                 TransformOrigin="@TransformOrigin"
>                 RelativeWidth="@RelativeWidth"
>                 Fixed="@DropdownSettings.Fixed"
>                 OverflowBehavior="@DropdownSettings.OverflowBehavior"
>                 DropShadow="@DropShadow"
>                 Duration="@(GetDense() ? 0 : MudGlobal.TransitionDefaults.Duration.TotalMilliseconds)">
>         <CascadingValue Value="@this">
>             <MudList T="object"
>                      Class="@ListClassname"
>                      Dense="@Dense"
>                      @onpointerenter="@PointerEnterAsync"
>                      @onpointerleave="@PointerLeaveAsync">
>                 @ChildContent
>             </MudList>
>         </CascadingValue>
>     </MudPopover>
> }
> 
> <MudOverlay Visible="_openState.Value && !_isTransient"
>             AutoClose
>             OnClosed="CloseMenuAsync"
>             LockScroll="LockScroll"
>             Modal="@Modal" />
> ```

Do it.

versile2 commented:
> Hey @danielchalmers, just a quick heads-up, the changes in #10551 seem to have broken the Nested Menu feature; I saw you were involved with it a bit.
> 
> While debugging and testing with child menus, I hit a big red `ObjectDisposedException`. Looks like the `CancellationTokenSource` fields accessed in `CancelPendingActions` are disposed but not set to `null`.
> 
> Setting them to `null` avoids the exception. However, I don't think that's the root cause, something else seems wrong, possibly a timing issue when the child menu appears and the mouse is moved slightly (triggering the render method).
> ### Steps to reproduce:
> 
>     1. Click **"OPEN NESTED MENU"**
> 
>     2. Hover over **"Format >"** so the child menu shows up
> 
>     3. Keep moving the mouse over **"Format >"**
> 
>     4. The child menu disappears and all hover effects stop working
> 
> 
> I can file a proper issue if needed, but trying to stay focused on this and avoid getting sidetracked too much 😆
> 
> https://mudblazor.com/components/menu#nested-menu-(desktop)

did you remove the surrounding if yet? That could cause it.

Cybrosys commented:
> > Hey @danielchalmers, just a quick heads-up, the changes in #10551 seem to have broken the Nested Menu feature; I saw you were involved with it a bit.
> > While debugging and testing with child menus, I hit a big red `ObjectDisposedException`. Looks like the `CancellationTokenSource` fields accessed in `CancelPendingActions` are disposed but not set to `null`.
> > Setting them to `null` avoids the exception. However, I don't think that's the root cause, something else seems wrong, possibly a timing issue when the child menu appears and the mouse is moved slightly (triggering the render method).
> > ### Steps to reproduce:
> > ```
> > 1. Click **"OPEN NESTED MENU"**
> > 
> > 2. Hover over **"Format >"** so the child menu shows up
> > 
> > 3. Keep moving the mouse over **"Format >"**
> > 
> > 4. The child menu disappears and all hover effects stop working
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I can file a proper issue if needed, but trying to stay focused on this and avoid getting sidetracked too much 😆
> > https://mudblazor.com/components/menu#nested-menu-(desktop)
> 
> did you remove the surrounding if yet? That could cause it.

Yes, I tried different variations. As it is in prod, after moving the overlay outside of it and removing the if completely. There are undesired behaviors and errors in all three configurations, so I think the changes broke something. It also depends on if you hover over the first menu or the child menu. Something triggers after a second or so and then it just breaks.

Cybrosys committed:
Add new PointerEventsNoneService, update Overlay, remove previous factory listener impl. and update unit tests

Cybrosys committed:
Fix ES6 compliance

Cybrosys committed:
Adjust javascript to satisfy SQ's wishes

Cybrosys committed:
Documented PointerEventsNoneObserver and sealed class PointerEventsObserverIgnore.

Cybrosys committed:
Documented PointerEventsNoneOptions and removed overlapping ctors

Cybrosys committed:
Updated documentation for MudOverlay and PointerEventsNonService

Cybrosys committed:
Documented IPointerEventsNoneObserver, IPointerDownObserver and IPointerUpObserver

danielchalmers commented:
> just a quick heads-up, the changes in #10551 seem to have broken the Nested Menu feature; I saw you were involved with it a bit.

Thanks! Should I wait until after this PR to fix that?

Cybrosys commented:
> > just a quick heads-up, the changes in #10551 seem to have broken the Nested Menu feature; I saw you were involved with it a bit.
> 
> Thanks! Should I wait until after this PR to fix that?

That probably depends on when @ScarletKuro has time to review my changes. If it’ll take a little while or if you know how to fix it then go ahead and I’ll resolve any merge conflicts afterward 😃 

@ScarletKuro I’ve removed the overlay-specific service and replaced it with a more general-purpose `IPointerEventsNoneService`, taking inspiration from `IKeyInterceptorService` and `IBrowserViewportService`, as you suggested. Let me know if this feels more in line with your expectations or how I should improve it.

The JavaScript portion is where I deviated slightly. it’s still close to what it was before. We could definitely refine it more, but I felt I needed to land on something that strikes a good balance for now.

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

ScarletKuro committed:
Nits

ScarletKuro commented:
Service looks good to me, I made it internal for now. Don't think we want to expose it to the end developer so we could add breaking change in future and this service looks too niche to be used widely by others?

henon commented:
> Service looks good to me, I made it internal for now. Don't think we want to expose it to the end developer so we could add breaking change in future and this service looks too niche to be used widely by others?

Agreed.

Cybrosys committed:
Add unit tests for PointerEventsNoneObserver

Cybrosys committed:
Add test for MudOverlay ElementId and rendered div ID

Cybrosys commented:
Target beat by 1.07% 💪 😆 - All green

Daniel pointed out earlier that there seems to be a behavior where clicking a `MudSelect` (and derived controls like `MudAutocomplete`) causes it to reopen immediately instead of closing, when `Modal` is `false` and it's clicked directly. I haven’t looked into a fix, but it appears related to the control relying on the overlay to handle closing.

Do you think this is something that could be addressed in this PR, or would it make more sense to handle it separately?

versile2 commented:
He fixed in another PR already I am pretty sure.

Cybrosys committed:
Add additional tests for PointerEventsNoneObserver

versile2 reviewed (approved):
Code looks good, tests look good., Thanks!

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

ScarletKuro commented:
@danielchalmers feel free to merge after done testing/reviewing.

danielchalmers commented:
Thanks, I'll look Tuesday. Fyi there was a new conflict with MudMenu.razor today.

Cybrosys committed:
Merge branch 'dev' into feature/mudmenu-modal-toggle

Cybrosys committed:
Fix bad merge

Cybrosys commented:
> Fyi there was a new conflict with MudMenu.razor today.

Resolved

danielchalmers reviewed (approved):
I don't see any problems aside from the behavior of some components which can be taken care of in another PR. E.g. Popovers should generally *toggle* if already visible, not always show (e.g. Autocomplete, Time Picker)

danielchalmers commented:
Great work! Looking forward to making this the default (v9?)

Alerinos commented:
It would be useful to have the same functionality in ``Dialog`` https://mudblazor.com/components/dialog#api

Cybrosys commented:
> It would be useful to have the same functionality in `Dialog` https://mudblazor.com/components/dialog#api

It was intentionally excluded because a `Dialog` is a true modal component. It should block interaction with the background and require a complete pointer down–up input sequence before dismissal. 

I believe the component `Popover` can give you the desired interaction, it does not prevent interaction with the background.

Alerinos commented:
@Cybrosys The dialogue lacks the message of the entire load when it is unexpectedly closed. Currently there is information that it is closed, without the payload, such as the model that fills the data in it. An option that would allow this would be useful.

ScarletKuro commented:
> @Cybrosys The dialogue lacks the message of the entire load when it is unexpectedly closed. Currently there is information that it is closed, without the payload, such as the model that fills the data in it. An option that would allow this would be useful.

Am I missing something, or how the `Modal` had anything to do with that

ScarletKuro commented:
> It was intentionally excluded because a `Dialog` is a true modal component. It should block interaction with the background and require a complete pointer down–up input sequence before dismissal.

The only way I see it as useful is if we look at Radzen, which allows opening the dialog at the top, bottom, left, or right, like a Flyout, and you can interact with the background. However, we don't have "Flyout" functionality in our dialog.

versile2 commented:
We should that's just cool. IMO :) Also great job on this PR.

henon commented:
We have flyout functionality in MudDrawer if I'm not mistaken

versile2 commented:
> We have flyout functionality in MudDrawer if I'm not mistaken

I've found it difficult at minimum to have two drawers at a time though. I need a todo to add that implementation to Docs, the examples show it but it doesn't work that easily in replication. I think another Doc item for "Fly Out" would be cool. I'll add it to my todo!

mwerberacs commented:
> Maybe add `Modeless` or `Modal` to `MudGlobal` (or some other place) so it's possible to opt-in globally for the new behavior?

I understand not making this default behavior until the next major release, but Is it possible for this to be added in the interim? Love this feature, but having to add the Modal attribute to all the relevant components individually is painful.

danielchalmers commented:
> > Maybe add `Modeless` or `Modal` to `MudGlobal` (or some other place) so it's possible to opt-in globally for the new behavior?
> 
> I understand not making this default behavior until the next major release, but is it possible for this to be added in the interim? Love this feature, but having to add the Modal attribute to all the relevant components individually is painful.

@mwerberacs Sure, I think we'd accept a PR for that if you'd like to submit one
